spawn_worker/src/unix/spawn.rs

```diff
@@ -0,0 +1,266 @@
+    let tmp_file = tempfile::NamedTempFile::new()?;
+    let mut file = tmp_file.as_file();
+    file.set_len(TRAMPOLINE_BIN.len() as u64)?;
+    file.write_all(TRAMPOLINE_BIN)?;
+    file.rewind()?;
+
+    std::fs::set_permissions(tmp_file.path(), Permissions::from_mode(0o700))?;
+
+    Ok(tmp_file)
+}
+
+pub enum SpawnMethod {
+    #[cfg(target_os = "linux")]
+    FdExec,
+    Exec,
+}
+
+impl Default for SpawnMethod {
+    #[cfg(target_os = "linux")]
+    fn default() -> Self {
+        Self::FdExec
+    }
+
+    #[cfg(not(target_os = "linux"))]
+    fn default() -> Self {
+        Self::Exec
+    }
+}
+
+pub enum Target {
+    Trampoline(extern "C" fn()),
+    ManualTrampoline(CString, CString),
+    Fork(fn()),
+    Noop,
+}
+
+pub struct SpawnCfg {
+    stdin: Option<OwnedFd>,
+    stderr: Option<OwnedFd>,
+    stdout: Option<OwnedFd>,
+    spawn_method: SpawnMethod,
+    target: Target,
+    inherit_env: bool,
+}
+
+impl SpawnCfg {
+    pub fn new() -> Self {
+        Self {
+            stdin: None,
+            stdout: None,
+            stderr: None,
+            target: Target::Noop,
+            inherit_env: true,
+            spawn_method: Default::default(),
+        }
+    }
+
+    pub fn target(&mut self, target: Target) -> &mut Self {
+        self.target = target;
+        self
+    }
+
+    pub fn stdin<T: Into<OwnedFd>>(&mut self, fd: T) -> &mut Self {
+        self.stdin = Some(fd.into());
+        self
+    }
+
+    pub fn stdout<T: Into<OwnedFd>>(&mut self, fd: T) -> &mut Self {
+        self.stdout = Some(fd.into());
+        self
+    }
+
+    pub fn stderr<T: Into<OwnedFd>>(&mut self, fd: T) -> &mut Self {
+        self.stderr = Some(fd.into());
+        self
+    }
+
+    pub fn spawn(&mut self) -> anyhow::Result<Child> {
+        let pid = self.do_spawn()?;
+
+        Ok(Child {
+            pid,
+            stdin: self.stdin.take(),
+            stderr: self.stderr.take(),
+            stdout: self.stdout.take(),
+        })
+    }
+
+    fn do_spawn(&self) -> anyhow::Result<Option<libc::pid_t>> {
+        let mut argv = ExecVec::default();
+        // set prog name (argv[0])
+        argv.push(CString::new("trampoline")?);
+        type SpawnFn = dyn Fn(&SealedExecVec, &SealedExecVec);
+
+        let spawn: Box<SpawnFn> = match &self.spawn_method {
+            #[cfg(target_os = "linux")]
+            SpawnMethod::FdExec => {
+                let fd = linux::write_trampoline()?;
+                Box::new(move |argv, envp| {
+                    // not using nix crate here, as it would allocate args after fork, which will lead to crashes on systems
+                    // where allocator is not fork+thread safe
+                    unsafe { libc::fexecve(fd.as_raw_fd(), argv.as_ptr(), envp.as_ptr()) };
+                    // if we're here then exec has failed
+                    panic!("{}", std::io::Error::last_os_error());
+                })
+            }
+            SpawnMethod::Exec => {
```

Comment:
what is the mechanism to know if we already did this write and that the daemon is already running ? I'm basically worried about short lived processes starting many instances.

Reply:
nope, this PR - only has the raw, spawn a subprocess - type functionality.

The logic about when the subprocess should be spawn - lives in a different PR